### PR TITLE
1459985: Handle job creation failure graciously

### PR DESCRIPTION
**Pull request by vritant**

### CODE REVIEW

**Comment by vritant** [comment]
[1460731, 1459985 Refresh Pools Jobs stuck in created state](https://trello.com/c/VCldq8YG/449-1460731-1459985-refresh-pools-jobs-stuck-in-created-state)

File: server/src/main/java/org/candlepin/pinsetter/tasks/KingpinJob.java [hunk 1]
```diff
@@ -152,20 +152,26 @@ public static JobStatus scheduleJob(JobCurator jobCurator,
...
-        if (trigger != null) {
-            scheduler.scheduleJob(detail, trigger);
-        }
-        else {
-            scheduler.addJob(detail, false);
+        catch (RuntimeException e) {
+            // if there was any error in scheduling, ensure that the status is updated
+            status.setState(JobStatus.JobState.FAILED);
+            jobCurator.merge(status);
+            throw e;
```
Comment (barnabycourt) [inline server/src/main/java/org/candlepin/pinsetter/tasks/KingpinJob.java]: Wasn't there an issue that throwing here not ending up in the log?

File: server/src/main/java/org/candlepin/pinsetter/tasks/KingpinJob.java [hunk 2]
```diff
@@ -152,20 +152,26 @@ public static JobStatus scheduleJob(JobCurator jobCurator,
...
             status = jobCurator.find(detail.getKey().getName());
             jobCurator.merge(status);
         }
-
-        if (trigger != null) {
-            scheduler.scheduleJob(detail, trigger);
-        }
-        else {
-            scheduler.addJob(detail, false);
+        catch (RuntimeException e) {
```
Comment (barnabycourt) [inline server/src/main/java/org/candlepin/pinsetter/tasks/KingpinJob.java]: Should this catch Throwable? In what exception case would the job not be marked as failed?

File: server/src/main/java/org/candlepin/pinsetter/tasks/KingpinJob.java (hunk 1 above)
Comment (vritant) [inline server/src/main/java/org/candlepin/pinsetter/tasks/KingpinJob.java]: spoke offline, we did and do continue to see stack traces.

File: server/src/main/java/org/candlepin/pinsetter/tasks/KingpinJob.java (hunk 2 above)
Comment (vritant) [inline server/src/main/java/org/candlepin/pinsetter/tasks/KingpinJob.java]: @barnabycourt the only exceptions we should need to catch are runtime exceptions imo.
it cannot not be possible to throw checked exceptions here, given that the method compiles without a throws clause.

Comment (vritant) [inline server/src/main/java/org/candlepin/pinsetter/tasks/KingpinJob.java]: oh wait there is a scheduler exception. looking deeper into that.

Comment (vritant) [inline server/src/main/java/org/candlepin/pinsetter/tasks/KingpinJob.java]: fixed

File: server/src/main/java/org/candlepin/pinsetter/tasks/KingpinJob.java [hunk 3]
```diff
@@ -152,23 +152,37 @@ public static JobStatus scheduleJob(JobCurator jobCurator,
...
             // status exists, let's update it
             // in theory this should be the rare case
             status = jobCurator.find(detail.getKey().getName());
             jobCurator.merge(status);
         }
-
-        if (trigger != null) {
-            scheduler.scheduleJob(detail, trigger);
+        catch (RuntimeException e) {
+            failStatus(jobCurator, status);
```
Comment (barnabycourt) [inline server/src/main/java/org/candlepin/pinsetter/tasks/KingpinJob.java]: Won't status be null jobCreator.create raises an exception?

**Comment by mstead** [comment]
This looks good to me. ACK.

**Comment by mstead** [comment]
retest this please

**Comment by vritant** [comment]
@candlepin-pull-request-bot test this please

**Comment by barnabycourt** [comment]
test this please

**Pull request merged**